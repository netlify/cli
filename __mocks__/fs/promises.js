// eslint-disable-next-line no-undef, @typescript-eslint/no-require-imports
const { fs } = require('memfs')

// eslint-disable-next-line no-undef
module.exports = fs.promises
