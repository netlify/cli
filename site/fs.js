'use strict'

const fs = require('fs').promises
const path = require('path')
const util = require('util')

const rimraf = require('rimraf')

const rimrafAsync = util.promisify(rimraf)

const copyDirRecursiveAsync = async (src, dest) => {
  try {
    fs.mkdir(dest, { recursive: true })
  } catch (_) {
    // ignore erros for mkdir
  }

  const childrenItems = await fs.readdir(src)

  await Promise.all(
    childrenItems.map(async (item) => {
      const srcPath = path.join(src, item)
      const destPath = path.join(dest, item)

      const itemStat = await fs.lstat(srcPath)

      if (itemStat.isFile()) {
        fs.copyFile(srcPath, destPath)
      } else {
        await copyDirRecursiveAsync(srcPath, destPath)
      }
    }),
  )
}

const ensureFilePathAsync = async (filePath) => {
  try {
    await fs.mkdir(path.dirname(filePath), { recursive: true })
  } catch (_) {
    // ignore any errors with mkdir - it will throw if the path already exists.
  }
}

const removeRecursiveAsync = async (filePath) => {
  await rimrafAsync(filePath)
}

module.exports = {
  copyDirRecursiveAsync,
  ensureFilePathAsync,
  removeRecursiveAsync,
}
