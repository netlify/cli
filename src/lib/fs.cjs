// @ts-check
const {
  constants,
  promises: { access, readFile, rm, stat },
} = require('fs')
const { version } = require('process')

const del = require('del')
const { gte, parse } = require('semver')

const NODE_VERSION = parse(version)

/**
 * reads a file async and catches potential errors
 * @param {string} filePath
 */
const readFileAsyncCatchError = async (filePath) => {
  try {
    return { content: await readFile(filePath, 'utf-8') }
  } catch (error) {
    return { error }
  }
}

const fileExistsAsync = async (filePath) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Removes a directory recursively and async
 * @param {string} path
 * @returns {Promise<void>}
 */
const rmdirRecursiveAsync = async (path) => {
  if (gte(NODE_VERSION, '14.14.0')) {
    return await rm(path, { force: true, recursive: true })
  }
  await del(path, { force: true })
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */
const isType = async (filePath, type) => {
  try {
    const stats = await stat(filePath)
    // @ts-ignore
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    if (error_.code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */
const isFileAsync = (filePath) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */
const isDirectoryAsync = (filePath) => isType(filePath, 'isDirectory')

module.exports = {
  fileExistsAsync,
  isDirectoryAsync,
  isFileAsync,
  readFileAsyncCatchError,
  rmdirRecursiveAsync,
}
