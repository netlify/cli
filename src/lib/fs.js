const fs = require('fs')
const pathType = require('path-type')
const { promisify } = require('util')

const statAsync = promisify(fs.stat)
const readFileAsync = promisify(fs.readFile)
const writeFileAsync = promisify(fs.writeFile)
const accessAsync = promisify(fs.access)

const readFileAsyncCatchError = async filepath => {
  try {
    return { content: await readFileAsync(filepath) }
  } catch (error) {
    return { error }
  }
}

const fileExistsAsync = async filePath => {
  try {
    await accessAsync(filePath, fs.F_OK)
    return true
  } catch (_) {
    return false
  }
}

const isFileAsync = filePath => pathType.isFile(filePath)

module.exports = { statAsync, readFileAsync, readFileAsyncCatchError, writeFileAsync, fileExistsAsync, isFileAsync }
