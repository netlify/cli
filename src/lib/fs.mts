// @ts-check
const {
  
  constants,
  
  promises: { access, readFile, rm, stat },
} = require('fs')

const { version } = require('process')

import del from 'del'

const { gte, parse } = require('semver')

const NODE_VERSION = parse(version)

/**
 * reads a file async and catches potential errors
 * @param {string} filePath
 */

const readFileAsyncCatchError = async (filePath: $TSFixMe) => {
  try {
    return { content: await readFile(filePath, 'utf-8') }
  } catch (error) {
    return { error }
  }
}


const fileExistsAsync = async (filePath: $TSFixMe) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Removes a directory recursively and async
 * @param {string} path
 * @returns {Promise<void>}
 */

const rmdirRecursiveAsync = async (path: $TSFixMe) => {
  if (gte(NODE_VERSION, '14.14.0')) {
    return await rm(path, { force: true, recursive: true })
  }
  await del(path, { force: true })
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */

const isType = async (filePath: $TSFixMe, type: $TSFixMe) => {
  try {
    const stats = await stat(filePath)
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    
    if ((error_ as $TSFixMe).code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */

const isFileAsync = (filePath: $TSFixMe) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */

const isDirectoryAsync = (filePath: $TSFixMe) => isType(filePath, 'isDirectory')

export default {
  fileExistsAsync,
  isDirectoryAsync,
  isFileAsync,
  readFileAsyncCatchError,
  rmdirRecursiveAsync,
}
