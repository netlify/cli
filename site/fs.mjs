import { promises as fs } from 'fs'
import { join, dirname } from 'path'
import { promisify } from 'util'

import rimraf from 'rimraf'

const rimrafAsync = promisify(rimraf)

export const copyDirRecursiveAsync = async (src, dest) => {
  try {
    fs.mkdir(dest, { recursive: true })
  } catch {
    // ignore erros for mkdir
  }

  const childrenItems = await fs.readdir(src)

  await Promise.all(
    childrenItems.map(async (item) => {
      const srcPath = join(src, item)
      const destPath = join(dest, item)

      const itemStat = await fs.lstat(srcPath)

      if (itemStat.isFile()) {
        fs.copyFile(srcPath, destPath)
      } else {
        await copyDirRecursiveAsync(srcPath, destPath)
      }
    }),
  )
}

export const ensureFilePathAsync = async (filePath) => {
  try {
    await fs.mkdir(dirname(filePath), { recursive: true })
  } catch {
    // ignore any errors with mkdir - it will throw if the path already exists.
  }
}

export const removeRecursiveAsync = async (filePath) => {
  await rimrafAsync(filePath)
}
