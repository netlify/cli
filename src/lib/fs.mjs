// @ts-check
import { constants } from 'fs'
import { access, stat } from 'fs/promises'

export const fileExistsAsync = async (filePath) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * calls stat async with a function and catches potential errors
 * @param {string} filePath
 * @param {keyof import('fs').StatsBase<number>} type For example `isDirectory` or `isFile`
 */
const isType = async (filePath, type) => {
  try {
    const stats = await stat(filePath)
    // @ts-ignore
    return typeof stats[type] === 'function' ? stats[type]() : stats[type]
  } catch (error_) {
    if (error_.code === 'ENOENT') {
      return false
    }

    throw error_
  }
}

/**
 * Checks if the provided filePath is a file
 * @param {string} filePath
 */
export const isFileAsync = (filePath) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 * @param {string} filePath
 */
export const isDirectoryAsync = (filePath) => isType(filePath, 'isDirectory')
