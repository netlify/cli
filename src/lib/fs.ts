import { constants } from 'fs'
import { access, stat } from 'fs/promises'

const isErrnoException = (value: unknown): value is NodeJS.ErrnoException =>
  value instanceof Error && Object.hasOwn(value, 'code')

export const fileExistsAsync = async (filePath: string) => {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

/**
 * calls stat async with a function and catches potential errors
 */
const isType = async (filePath: string, type: 'isFile' | 'isDirectory') => {
  try {
    const stats = await stat(filePath)
    if (type === 'isFile') return stats.isFile()
    return stats.isDirectory()
  } catch (error) {
    if (isErrnoException(error) && error.code === 'ENOENT') {
      return false
    }

    throw error
  }
}

/**
 * Checks if the provided filePath is a file
 */
export const isFileAsync = async (filePath: string) => isType(filePath, 'isFile')

/**
 * Checks if the provided filePath is a directory
 */
export const isDirectoryAsync = async (filePath: string) => isType(filePath, 'isDirectory')
