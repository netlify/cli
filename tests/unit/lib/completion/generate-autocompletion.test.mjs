// @ts-check
import fs from 'fs'

import test from 'ava'
import { Argument } from 'commander'
import sinon from 'sinon'

const createTestCommand = async () => {
  const { default: BaseCommand } = await import('../../../../src/commands/base-command.mjs')
  const program = new BaseCommand('chef')

  program
    .command('bake')
    .alias('cook')
    .description('Cooks something')
    .addExamples(['chef cook pizza'])
    .argument('<type>', 'the type to cook')
    .option('-f, --fast', 'cook it fast')

  program.command('bake:pizza').description('bakes a pizza').option('--type <type>', 'Type of pizza', 'neapolitan')

  program
    .command('taste')
    .description('tastes something')
    .addArgument(new Argument('<name>', 'what to taste').choices(['pizza', 'sauce']))

  return program
}

test.afterEach(() => {
  // eslint-disable-next-line import/no-named-as-default-member
  sinon.restore()
})

test('should generate a completion file', async (t) => {
  // eslint-disable-next-line import/no-named-as-default-member
  const stub = sinon.stub(fs, 'writeFileSync').callsFake(() => {})
  const { default: generateAutocompletion } = await import('../../../../src/lib/completion/generate-autocompletion.mjs')
  const program = await createTestCommand()

  generateAutocompletion(program)

  // @ts-ignore
  t.true(stub.getCall(0).args[0].endsWith('autocompletion.json'), 'should write a autocompletion file')

  // @ts-ignore
  t.snapshot(JSON.parse(stub.getCall(0).args[1]))

  stub.restore()
})
