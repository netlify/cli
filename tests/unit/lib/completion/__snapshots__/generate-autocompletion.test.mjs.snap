// Vitest Snapshot v1

exports[`generateAutocompletion > should generate a completion file 1`] = `undefined`;
